File: Templates/Segtree.java
class Segtree {
    public int n;
    public long[] tree;
    public long cmb(long a, long b) { return Math.min(a, b); }
    public final long id = Long.MAX_VALUE;
    public Segtree(int n) {
        this.n = 1;
        while (this.n < n) this.n <<= 1;
        tree = new long[this.n << 1];
        Arrays.fill(tree, id);
    }
    public void upd(int i, long x) {
        i += n;
        tree[i] = x; i >>= 1;
        for (; i > 0; i >>= 1) tree[i] = cmb(tree[i << 1], tree[i << 1 ^ 1]);
    }
    public long qry(int l, int r) {
        l += n; r += n;
        long res = id;
        while (l <= r) {
            if ((l & 1) == 1) res = cmb(res, tree[l++]);
            if ((r & 1) == 0) res = cmb(res, tree[r--]);
            l >>= 1; r >>= 1;
        }
        return res;
    }
}

File: Templates/Segtree.cpp
struct Segtree {
    int n;
    using T = int;
    const T id = INT_MAX;
    T cmb(T i, T j) { return min(i, j); }
    V<T> tree;
    Segtree(int _n) {
        n = 1;
        while (n < _n) n *= 2;
        tree.assign(2*n, id);
    }
    void upd(int i, T v) {
        i += n; tree[i] = v;
        while (i > 1) {
            i /= 2; tree[i] = cmb(tree[2*i], tree[2*i+1]);
        }
    }
    T qry(int i, int j) {
        T ra = id, rb = id;
        i += n, j += n;
        while (i <= j) {
            if (i % 2 == 1) ra = cmb(ra, tree[i++]);
            if (j % 2 == 0) rb = cmb(tree[j--], rb);
            i /= 2, j /= 2;
        }
        return cmb(ra, rb);
    }
};
